In [97]:
import numpy as np
A = np.array([[3,-1,-1],[-1.,3.,1.],[2,1,4]])
b = np.array([1.,3.,7.])



In [98]:
def SOR(A,b,x0,omega,itmax=10000,tolerancia=1e-14):
    
    x = x0.copy()
    u = x.copy()
    sumk = x.copy()
    
    residuo = np.linalg.norm( np.dot(A,x) - b)
    
    
    
    it = 0
    
    while residuo >= tolerancia and it < itmax:
        
        
        u[:] = 0
        sumk[:] = 0.
        
        for i in range(A.shape[0]):
            for j in range(A.shape[1]):
                if i != j:
                    sumk[i] += A[i,j]*x[j]
            #print(sumk)
        
            u[i] = (1-omega)*x[i]+(b[i] - sumk[i]-sumk[i-1])*(omega/A[i,i])
        
        
        x = u.copy()
        
        #print(x,residuo)
        # Norma infinita
        residuo = np.max(np.abs(np.dot(A,x) - b))
        # Norma griega
        #residuo = np.linalg.norm( np.dot(A,x) - b)
        
        it += 1
        
        if residuo > 1000:
            print('No calculable con Jacobi')
            x[:] = 0.
            break
        
    return x,it

In [99]:
om=np.linspace(1,2,101)
omega=1.06
r=SOR(A,b,np.array([0,0,0]), omega)
r


(array([1, 1, 1]), 2)

In [100]:
omegas=[]
its=[]
for o in om:
    x= SOR(A,b,np.array([0,0,0]), o)
    omegas.append(o)
    its.append(x[1])
print(its)
print(omegas)
it_min=min(its)
pos=its.index(it_min)
omega_min=omegas[pos]
omega_min


No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
No calculable con Jacobi
[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 10000, 9, 10000, 10000, 10000, 10000, 10000, 26, 10000, 54, 54, 43, 41, 41, 39, 27, 27, 27, 24, 24, 23, 22, 21, 20, 19, 19, 19, 16]
[1.0, 1.01, 1.02, 1.03, 1.04, 1.05, 1.06, 1.07, 1.08, 1.09, 1.1, 1.11, 1.12, 1.13, 1.1400000000000001, 1.15, 1.16, 1.17, 

1.0

Podemos observar que para 1< omega < 1.14 las iteraciones son unicamente 2, cualquiera de estos valores minimiza las iteraciones